spring-cloud-gcp-pubsub-stream-binder/src/test/java/org/springframework/cloud/gcp/stream/binder/pubsub/PubSubEmulator.java

```diff
@@ -0,0 +1,239 @@
+
+	private static final Path EMULATOR_CONFIG_PATH = EMULATOR_CONFIG_DIR.resolve(ENV_FILE_NAME);
+
+	private static final Log LOGGER = LogFactory.getLog(PubSubEmulator.class);
+
+	// Reference to emulator instance, for cleanup.
+	private Process emulatorProcess;
+
+	// Hostname for cleanup, should always be localhost.
+	private String emulatorHostPort;
+
+	// Conditional rule execution based on an environmental flag.
+	private boolean enableTests;
+
+	public PubSubEmulator() {
+		if ("true".equals(System.getProperty("it.pubsub-emulator"))) {
+			this.enableTests = true;
+		}
+		else {
+			LOGGER.warn("PubSubEmulator rule disabled. Please enable with -Dit.pubsub-emulator.");
+		}
+	}
+
+	/**
+	 * Launch an instance of pubsub emulator or skip all tests.
+	 * If it.pubsub-emulator environmental property is off, all tests will be skipped through the failed assumption.
+	 * If the property is on, any setup failure will trigger test failure. Failures during teardown are merely logged.
+	 * @throws IOException if config file creation or directory watcher on existing file fails.
+	 * @throws InterruptedException if process is stopped while waiting to retry.
+	 */
+	@Override
+	protected void before() throws IOException, InterruptedException {
+
+		assumeTrue("PubSubEmulator rule disabled. Please enable with -Dit.pubsub-emulator.", this.enableTests);
+
+		startEmulator();
+		determineHostPort();
+	}
+
+	/**
+	 * Shut down the two emulator processes.
+	 * gcloud command is shut down through the direct process handle. java process is
+	 * identified and shut down through shell commands.
+	 * There should normally be only one process with that host/port combination, but if there
+	 * are more, they will be cleaned up as well.
+	 * Any failure is logged and ignored since it's not critical to the tests' operation.
+	 */
+	@Override
+	protected void after() {
+		if (this.emulatorProcess == null) {
+			LOGGER.warn("Emulator process null after tests; nothing to terminate.");
+			return;
+		}
+
+		this.emulatorProcess.destroy();
+
+		if (this.emulatorHostPort == null) {
+			LOGGER.warn("Host/port null after the test.");
+			return;
+		}
+
+		try {
+			int portSeparatorIndex = this.emulatorHostPort.indexOf(":");
+			if (portSeparatorIndex < 0 || !this.emulatorHostPort.contains("localhost")) {
+				LOGGER.warn("Malformed host: " + this.emulatorHostPort);
+				return;
+			}
+			String emulatorHost = this.emulatorHostPort.substring(0, portSeparatorIndex);
+			String emulatorPort = this.emulatorHostPort.substring(portSeparatorIndex + 1);
+
+			String hostPortParams = String.format("--host=%s --port=%s", emulatorHost, emulatorPort);
+			Process psProcess = new ProcessBuilder("ps", "-v").start();
+			new BufferedReader(new InputStreamReader(psProcess.getInputStream()))
+					.lines()
+					.filter(psLine -> psLine.contains(hostPortParams))
+					.map(psLine -> new StringTokenizer(psLine).nextToken())
+					.forEach(this::killProcess);
```

Comment:
Don't we need to close the `Stream` after using it?
I believe it is going to close all the `BufferedReader` and `InputStream` it is based on...